### PR TITLE
[TypeSpec Validation] tsv folder path standardization

### LINKED ISSUE #26546: [TypeSpec Validation] "tsv ." fails with path error
On Ubuntu 22:

```
mharder@mharder-dev-u22:~/specs/specification/vision/ImageAnalysis$ npx tsv .
Running TypeSpecValidation on folder: .

Executing rule: FolderStructure
file:///home/mharder/specs/eng/tools/TypeSpecValidation/dist/src/rules/folder-structure.js:40
            if (folderStruct[1].match(/[A-Z]/g)) {
                                ^

TypeError: Cannot read properties of undefined (reading 'match')
    at FolderStructureRule.<anonymous> (file:///home/mharder/specs/eng/tools/TypeSpecValidation/dist/src/rules/folder-structure.js:40:33)
    at Generator.next (<anonymous>)
    at fulfilled (file:///home/mharder/specs/eng/tools/TypeSpecValidation/dist/src/rules/folder-structure.js:4:58)

Node.js v18.18.2
```

### PR DESCRIPTION
**Pull request by ckairen**
closes #26546

### CODE REVIEW

**Comment by ckairen** [comment]
tsv . will only work when it's in the correct spec folder. Since the tsv tool is intended to run on spec folders and not finding specs recursivley.

File: eng/tools/typespec-validation/src/rules/folder-structure.ts [hunk 1]
```diff
@@ -11,6 +11,8 @@ export class FolderStructureRule implements Rule {
     let success = true;
     let stdOutput = "";
     let errorOutput = "";
+    let gitRoot = path.normalize(await host.gitOperation(folder).revparse("--show-toplevel"));
+    let relativePath = path.relative(gitRoot, folder).split(path.sep).join("/");
```
Comment (mikeharder) [inline eng/tools/typespec-validation/src/rules/folder-structure.ts:15]: Add a unit test exercising these new APIs.

File: eng/tools/typespec-validation/src/index.ts [hunk 2]
```diff
@@ -18,8 +18,10 @@ export async function main() {
     },
   };
   const parsedArgs = parseArgs({ args, options, allowPositionals: true } as ParseArgsConfig);
-  const folder = parsedArgs.positionals[0].split(path.sep).join("/");
-  console.log("Running TypeSpecValidation on folder:", folder);
+  const folder = parsedArgs.positionals[0];
+  const absolutePath = path.resolve(folder).split(path.sep).join("/");
```
Comment (mikeharder) [inline eng/tools/typespec-validation/src/index.ts:22]: Extract a helper function named something like `normalizePath([...paths])` that can be unit tested, and add unit tests.  Unit tests should not rely on current directory.

**Review by mikeharder — CHANGES_REQUESTED** [review]

File: eng/tools/typespec-validation/test/folder-structure.test.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
+import { FolderStructureRule } from "../src/rules/folder-structure.js";
+import { TsvTestHost } from "./tsv-test-host.js";
+import { strict as assert } from "node:assert";
+
+describe("folder-structure", function () {
+  it("should fail if no tspconfig has incorrect extension", async function () {
```
Comment (mikeharder) [inline eng/tools/typespec-validation/test/folder-structure.test.ts:6]:
```suggestion
  it("should fail if tspconfig has incorrect extension", async function () {
```

Was this a typo where the intention was reversed?

File: eng/tools/typespec-validation/test/folder-structure.test.ts [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    host.globby = async () => {
+      return ["/foo/bar/tspconfig.yml"];
+    };
+
+    const result = await new FolderStructureRule().execute(host, TsvTestHost.folder);
+    assert(result.errorOutput);
+    assert(result.errorOutput.includes("Invalid config file"));
+  });
+
+  it("should fail if no folder under specification/ is capitalized", async function () {
```
Comment (mikeharder) [inline eng/tools/typespec-validation/test/folder-structure.test.ts:17]:
```suggestion
  it("should fail if folder under specification/ is capitalized", async function () {
```

File: eng/tools/typespec-validation/test/folder-structure.test.ts [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    host.globby = async () => {
+      return ["/foo/bar/tspconfig.yaml"];
+    };
+    host.normalizePath = () => {
+      return "/gitroot";
+    };
+
+    const result = await new FolderStructureRule().execute(
+      host,
+      "/gitroot/specification/Foo/Foo/Foo/Foo",
```
Comment (mikeharder) [inline eng/tools/typespec-validation/test/folder-structure.test.ts:42]:
```suggestion
      "/gitroot/specification/foo/Foo/Foo/Foo",
```

I think the service-family folder should be lower-case, unless we are trying to make it fail?

File: eng/tools/typespec-validation/test/folder-structure.test.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    host.globby = async () => {
+      return ["/foo/bar/tspconfig.yaml"];
+    };
+    host.normalizePath = () => {
+      return "/gitroot";
+    };
+
+    const result = await new FolderStructureRule().execute(
+      host,
+      "/gitroot/specification/Foo/Foo.foo",
```
Comment (mikeharder) [inline eng/tools/typespec-validation/test/folder-structure.test.ts:59]:
```suggestion
      "/gitroot/specification/foo/Foo.foo",
```

File: eng/tools/typespec-validation/test/folder-structure.test.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    host.globby = async () => {
+      return ["/foo/bar/tspconfig.yaml"];
+    };
+    host.normalizePath = () => {
+      return "/gitroot";
+    };
+
+    const result = await new FolderStructureRule().execute(
+      host,
+      "/gitroot/specification/Foo/Foo.Management.Foo.Shared",
```
Comment (mikeharder) [inline eng/tools/typespec-validation/test/folder-structure.test.ts:76]:
```suggestion
      "/gitroot/specification/foo/Foo.Management.Foo.Shared",
```

File: eng/tools/typespec-validation/test/folder-structure.test.ts [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        return false;
+      } else if (file.includes("client.tsp")) {
+        return false;
+      }
+      return true;
+    };
+
+    const result = await new FolderStructureRule().execute(
+      host,
+      "/gitroot/specification/Foo/Foo.Management",
```
Comment (mikeharder) [inline eng/tools/typespec-validation/test/folder-structure.test.ts:101]:
```suggestion
      "/gitroot/specification/foo/Foo.Management",
```

File: eng/tools/typespec-validation/test/folder-structure.test.ts [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        return true;
+      } else if (file.includes("examples")) {
+        return false;
+      }
+      return true;
+    };
+
+    const result = await new FolderStructureRule().execute(
+      host,
+      "/gitroot/specification/Foo/Foo.Management",
```
Comment (mikeharder) [inline eng/tools/typespec-validation/test/folder-structure.test.ts:127]:
```suggestion
      "/gitroot/specification/foo/Foo.Management",
```

File: eng/tools/typespec-validation/test/folder-structure.test.ts [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    host.checkFileExists = async (file: string) => {
+      if (file.includes("tspconfig.yaml")) {
+        return false;
+      }
+      return true;
+    };
+
+    const result = await new FolderStructureRule().execute(
+      host,
+      "/gitroot/specification/Foo/Foo.Management",
```
Comment (mikeharder) [inline eng/tools/typespec-validation/test/folder-structure.test.ts:151]:
```suggestion
      "/gitroot/specification/foo/Foo.Management",
```

**Review by mikeharder — APPROVED** [review]
Approved pending fixes to tests.

File: eng/tools/typespec-validation/src/index.ts [hunk 11]
```diff
@@ -18,10 +18,10 @@ export async function main() {
     },
   };
   const parsedArgs = parseArgs({ args, options, allowPositionals: true } as ParseArgsConfig);
-  const folder = parsedArgs.positionals[0].split(path.sep).join("/");
-  console.log("Running TypeSpecValidation on folder:", folder);
+  const folder = parsedArgs.positionals[0];
+  const absolutePath = host.normalizePath(folder);
 
-  const host = new TsvRunnerHost();
+  console.log("Running TypeSpecValidation on folder:", absolutePath);
```
Comment (konrad-jamrozik) [inline eng/tools/typespec-validation/src/index.ts:24]: @ckairen I think you may want to add a space after `folder:`

Comment (ckairen) [inline eng/tools/typespec-validation/src/index.ts:24]:
```suggestion
  console.log("Running TypeSpecValidation on folder: ", absolutePath);
```

**Pull request merged**